flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/FsSavepointStore.java

```diff
@@ -0,0 +1,225 @@
+			// Write header
+			dos.writeInt(MAGIC_NUMBER);
+			dos.writeInt(savepoint.getVersion());
+
+			// Write savepoint
+			SavepointSerializer<T> serializer = SavepointSerializers.getSerializer(savepoint);
+			serializer.serialize(savepoint, fdos);
+			success = true;
+		} finally {
+			fdos.close();
+
+			if (!success && fileSystem.exists(path)) {
+				if (!fileSystem.delete(path, true)) {
+					LOG.warn("Failed to delete file " + path + " after failed write.");
+				}
+			}
+		}
+
+		return path.toString();
+	}
+
+	@Override
+	public Savepoint loadSavepoint(String path) throws IOException {
+		Preconditions.checkNotNull(path, "Path");
+
+		try {
+			try (FSDataInputStream fdis = createFsInputStream(new Path(path))) {
+				try (DataInputStream dis = new DataInputStream(fdis)) {
+					int magicNumber = dis.readInt();
+					int version = dis.readInt();
+
+					if (magicNumber == MAGIC_NUMBER) {
+						SavepointSerializer<?> serializer = SavepointSerializers.getSerializer(version);
+						Savepoint savepoint = serializer.deserialize(fdis);
+						return savepoint;
+					} else {
+						throw new IllegalStateException("Unexpected magic number. This indicates " +
+								"that the specified file is not a proper savepoint or " +
+								"the file has been corrupted.");
+					}
+				}
+			}
+		} catch (Throwable t) {
+			// Flink 1.0 did not store a header, check if it is an old savepoint.
+			// Only after this fails as well we can be sure that it is an actual
+			// failure.
+			Savepoint savepoint = tryLoadSavepointV0(path);
+			if (savepoint == null) {
+				throw new IOException("Failed to load savepoint " + path + ".", t);
+			} else {
+				return savepoint;
+			}
+		}
+	}
+
+	@Override
+	public void disposeSavepoint(String path, ClassLoader classLoader) throws Exception {
+		Preconditions.checkNotNull(path, "Path");
+		Preconditions.checkNotNull(classLoader, "Class loader");
+
+		try {
+			Savepoint savepoint = loadSavepoint(path);
+			savepoint.dispose(classLoader);
+
+			Path filePath = new Path(path);
+
+			if (fileSystem.exists(filePath)) {
+				if (!fileSystem.delete(filePath, true)) {
+					throw new IOException("Failed to delete " + filePath + ".");
+				}
+			} else {
+				throw new IllegalArgumentException("Invalid path '" + filePath.toUri() + "'.");
+			}
+		} catch (Throwable t) {
+			throw new IOException("Failed to dispose savepoint " + path + ".", t);
+		}
+	}
+
+	@Override
+	public void shutDown() throws Exception {
+		// Nothing to do, because the savepoint life-cycle is independent of
+		// the cluster life-cycle.
+	}
+
+	private FSDataInputStream createFsInputStream(Path path) throws IOException {
+		if (fileSystem.exists(path)) {
+			return fileSystem.open(path);
+		} else {
+			throw new IllegalArgumentException("Invalid path '" + path.toUri() + "'.");
+		}
+	}
+
+	/**
+	 * Tries to load a Flink 1.0 savepoint, which was not stored with a header.
+	 *
+	 * <p>This can be removed for Flink 1.2.
```

Comment:
Why can we remove this for Flink 1.2? Wouldn't that mean that Flink 1.2 can only restore savepoints from version 1.1 onwards? I think it should also be possible to restore savepoints from version 1.0 with version 1.2.